modules/dnn/src/opencl/ocl4dnn_lrn.cl

```diff
@@ -0,0 +1,604 @@
+      accum_ratio += top_diff_off[head * step] * top_off[head * step]
+          / scale_off[head * step];
+      if (head - size >= 0) {
+        accum_ratio -= top_diff_off[(head - size) * step]
+            * top_off[(head - size) * step] / scale_off[(head - size) * step];
+      }
+      bottom_diff_off[(head - post_pad) * step] = top_diff_off[(head - post_pad)
+          * step] * pow(scale_off[(head - post_pad) * step], negative_beta)
+          - cache_ratio * bottom_off[(head - post_pad) * step] * accum_ratio;
+      ++head;
+    }
+    // subtract only
+    while (head < channels + post_pad) {
+      if (head - size >= 0) {
+        accum_ratio -= top_diff_off[(head - size) * step]
+            * top_off[(head - size) * step] / scale_off[(head - size) * step];
+      }
+      bottom_diff_off[(head - post_pad) * step] = top_diff_off[(head - post_pad)
+          * step] * pow(scale_off[(head - post_pad) * step], negative_beta)
+          - cache_ratio * bottom_off[(head - post_pad) * step] * accum_ratio;
+      ++head;
+    }
+  }
+}
+
+// This kernel is used in LRNOptimizedLayer.
+// It is based on ave_pool_forward kernel with some modifications on the input and output.
+__kernel void TEMPLATE(lrn_within_channel_forward,Dtype)(
+    const int nthreads, __global const Dtype* const bottom_data, const int num,
+    const int channels, const int height, const int width,
+    const int pooled_height, const int pooled_width, const int kernel_h,
+    const int kernel_w, const int stride_h, const int stride_w, const int pad_h,
+    const int pad_w, const Dtype power, const Dtype scale, __global Dtype* top_data) {
+  for (int index = get_global_id(0); index < nthreads;
+      index += get_global_size(0)) {
+    {
+      const int pw = index % pooled_width;
+      const int ph = (index / pooled_width) % pooled_height;
+      const int c = (index / pooled_width / pooled_height) % channels;
+      const int n = index / pooled_width / pooled_height / channels;
+      int hstart = ph * stride_h - pad_h;
+      int wstart = pw * stride_w - pad_w;
+      int hend = min(hstart + kernel_h, height + pad_h);
+      int wend = min(wstart + kernel_w, width + pad_w);
+      const int pool_size = (hend - hstart) * (wend - wstart);
+      hstart = max(hstart, (int)0);
+      wstart = max(wstart, (int)0);
+      hend = min(hend, height);
+      wend = min(wend, width);
+      Dtype aveval = 0;
+      __global const Dtype* bottom_slice = bottom_data
+          + (n * channels + c) * height * width;
+      for (int h = hstart; h < hend; ++h) {
+        for (int w = wstart; w < wend; ++w) {
+          Dtype tmp_val = bottom_slice[h * width + w];
+          aveval += (tmp_val * tmp_val);
+        }
+      }
+          Dtype tmp_aveval = aveval / pool_size;
+          top_data[index] = bottom_data[index] * pow(tmp_aveval * scale + 1, power);
+    }
+  }
+}
+
+
+// !!!!!!!!
+// The following macros are duplicated in lrn_optimized_layer.cu since we currently
```

Comment:
Probably this comment is not relevant anymore

Reply:
comments are cleaned.